Handle hitless events post Sophronia

gonzaponte opened the pull request:
As discussed in #897, we implement a solution to hitless events in Esmeralda and Beersheba. Hitless events are preserved in the `/Run/events` table but are (naturally) missing in `/RECO/Events`.

Closes #897.

gonzaponte committed:
Add test to demonstrate the issue

gonzaponte committed:
Remove comparison between hits and event table

gonzaponte committed:
Warn about hitless events

gonzaponte committed:
Extend test to show new behaviour

jwaiton commented on invisible_cities/cities/components.py:
```diff
@@ -594,12 +594,17 @@ def hits_and_kdst_from_files( paths : List[str]
                 event_number, timestamp = evtinfo.fetch_all_fields()
                 hits = hits_from_df(hits_df.loc[hits_df.event == event_number])
-                yield dict(hits = hits[event_number],
+                if len(hits):
+                    hits = hits[event_number]
+                else:
+                    warnings.warn(f"Event {event_number} does not contain hits", UserWarning)
+                    hits = HitCollection(event_number, timestamp, [])
+                yield dict(hits = hits,
```
`if len(hits):` returns `True` for any number of rows in the `hits` dataframe, and `False` only for 0 rows, is that correct? This usage of `if len(hits):` interpreting the length of a dataframe like this feels a bit obtuse to me, but perhaps that is due to my lack of familiarity with boolean behaviour in python.

gonzaponte replied:
Yes, this is idiomatic in Python. When interpreted as booleans, all numbers except 0 are `True` (even `np.nan`), and 0 is `False`.
This pattern is particularly common in sequences. `if len(something)` means "if it's not empty". You may also see `if not len(something)`, which means "if it's empty".
In this particular case, `hits` is a dictionary that maps event numbers to `HitCollection`s. The `len` of a dictionary corresponds to the number of entries and therefore the meaning above is preserved.

jacg replied:
Idiomatic Python would omit the `len`. Empty containers (be they sequences, mappings, whatever) are false in boolean contexts in Python and you are expected to take advantage of this ... except that numpy arrays throw a spanner in the works by being sequences in most respects, with the notable exception of their boolean interpretation. Consequently, in environments where numpy is used, it becomes less clear that the standard Python idioms are a good idea:

```python
import numpy as np

samples = (
                 ''  ,              'a'  ,
                 []  ,              [0]  ,
                 ()  ,              (0,) ,
                 {}  ,              {0:0},
              set()  ,              {0}  ,
    #   np.array([]) ,     np.array([0]) , # Oh dear, numpy is a git
    len(np.array([])), len(np.array([0])), # `len` provides a workaround
)

def bbb(x):
	print(x, bool(x))
    
for x in samples:
	bbb(x)
```

So, while the superfluous-for-all-containers-except-numpy-arrays `len`, is bad style in Python in general, it might be justified in projects that use numpy, for the sake of consistency and avoiding numpy-caused surprises.

jwaiton commented on invisible_cities/cities/sophronia_test.py:
```diff
@@ -105,3 +109,7 @@ def test_sophronia_filters_events_with_only_nn_hits(config_tmpdir, sophronia_con
+
+        # Check that it's present in /Run/events, but not in /RECO/Events
+        assert event_number == output_file.root.Run.events[0][0]
+        assert event_number not in output_file.root.RECO.Events.col("event")
```
This addition to `test_sophronia_filters_events_with_only_nn_hits()` covers the case we're considering (hitless events retaining `Run/events` information but not `RECO/Events` information), but the test itself isn't descriptive of this condition we're checking for outside of the comment L113. 

Is it worth separating this case we're checking for (hitless events showing appropriate behaviour) into a separate test to make clear that this is a condition we're checking for? Or perhaps improving the docstring/name of the test to ensure that this is clear?

gonzaponte committed:
Split test into two parts for clarity

jwaiton commented on invisible_cities/cities/esmeralda_test.py:
```diff
@@ -199,3 +200,22 @@ def test_esmeralda_filters_events_with_too_many_hits(esmeralda_config, Th228_tra
+        n_delete  = sum(evt_rows)
+        file.root.RECO.Events.remove_rows(0, n_delete)
+
+    # just test that it doesn't crash
+    esmeralda(**esmeralda_config)
```
I'm assuming that Beersheba works identically to this, and so it doesn't need a test to verify it doesn't crash?

gonzaponte replied:
It works in a similar fashion, but a test wouldn't hurt. This is also true for the component itself. I will add both.

gonzaponte committed:
Replicate hitless test for Beersheba

gonzaponte committed:
Fix `HitCollection` to DataFrame conversion

In the case of a hitless event, the output DataFrame did not contain
any columns, which caused problems later on

gonzaponte committed:
Add test for `hits_and_kdst_from_files` using hitless events

gonzaponte committed:
Refactor tests using fixture to provide file with missing hits

gonzaponte committed:
Filter warnings

jwaiton reviewed (approved):
Introduces a solution for hitless events post-Sophronia that stops Esmeralda and Beersheba from crashing, and provides a warning to the user if they occur. Good work!

gonzaponte committed:
Create temporary file in temporary folder
